File: a1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

const int MAX_NUM = 100000;

struct Value {
	int leng;
	int index;

    Value(int leng, int index) : leng(leng), index(index) {}

    Value() : leng(0), index(0) {}
};

vector<int> input() {
    string text;
	getline(cin, text);
	
	vector<int> arr;
    string temp = "";
    for (auto c: text) {
        if (c == ' ') {
            if (temp != "") arr.push_back(stoi(temp));
            temp = "";
        } else {
            temp += c;
        }
    }
    arr.push_back(stoi(temp));

	return arr;
}

void update(vector<vector<Value>>& fenwick_tree, int index, Value val) {
    for (int i = index; i <= MAX_NUM; i += i & -i) {
        int current_leng;
        if (fenwick_tree[i].size() > 0)
            current_leng = fenwick_tree[i][0].leng;
        else
            current_leng = 0;

        if (current_leng < val.leng) {
            fenwick_tree[i].clear();
            fenwick_tree[i].push_back(val);
        } else if (current_leng == val.leng)
            fenwick_tree[i].push_back(val);
    }
}

vector<Value> get(vector<vector<Value>>& fenwick_tree, int index) {
    vector<Value> res;
    int bestLeng = 0;
    for (int i = index; i > 0; i -= i & -i) 
        if (fenwick_tree[i].size() > 0) 
            if (bestLeng < fenwick_tree[i][0].leng)
                res = fenwick_tree[i], bestLeng = fenwick_tree[i][0].leng;
            else if (bestLeng == fenwick_tree[i][0].leng)
                res.insert(res.end(), fenwick_tree[i].begin(), fenwick_tree[i].end());
    return res;
}

void printAns(vector<int>& arr, vector<vector<Value>>& trace, int index, vector<int>& pocket, unordered_set<string>& list_output) {
    pocket.push_back(arr[index]);
    if (trace[index].size() != 0) {
        for (auto t: trace[index])
            printAns(arr, trace, t.index, pocket, list_output);
    } else {
        string temp = "";
        for (int i = pocket.size()-1; i >= 0; i--)
            temp += to_string(pocket[i]) + " ";
        list_output.insert(temp);
    }
    pocket.pop_back();
}

int main() {
    vector<Value> ans;
    int bestLeng = 0;
    vector<int> arr = input();

    vector<vector<Value>> fenwick_tree(MAX_NUM);
    vector<vector<Value>> trace(arr.size());
    for (int index = 0; index < arr.size(); index++) {
        auto val = get(fenwick_tree, arr[index]-1);
        trace[index] = val;
        
        int leng = 0;
        if (val.size() > 0) leng = val[0].leng;
        Value new_val(leng+1, index);

        update(fenwick_tree, arr[index], new_val);

        if (bestLeng < new_val.leng) {
            ans.clear();
            ans.push_back(new_val);
            bestLeng = new_val.leng;
        } else if (bestLeng == new_val.leng) 
            ans.push_back(new_val);
    }

    vector<int> pocket;
    unordered_set<string> list_output;
    for (auto x: ans)
        printAns(arr, trace, x.index, pocket, list_output);
    for (string s: list_output)
        cout << s << endl;
}

File: a1.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"strconv"
)

const MAX_NUM = 100000

type Value struct {
	leng int
	index int
}

func input() []int {
	reader := bufio.NewReader(os.Stdin)
	text, _ := reader.ReadString('\n')
	
	var arr []int
	for _, i := range strings.Split(text[:len(text)-1], " ") {
		if i != "" {
			num, err := strconv.Atoi(i)
			if err != nil {
				panic(err)
			}
			arr = append(arr, num)
		}
	}

	return arr
}

func update(fenwick_tree []Value, index int, val Value) {
	for i := index; i <= MAX_NUM; i += i & -i {
		if fenwick_tree[i].leng < val.leng {
			fenwick_tree[i] = val
		}
	}
}

func get(fenwick_tree []Value, index int) Value {
	var res Value
	for i := index; i > 0; i -= i & -i {
		if res.leng < fenwick_tree[i].leng {
			res = fenwick_tree[i]
		}
	}
	return res
}

func printAns(arr []int, trace []Value, index int) {
	if trace[index].leng != 0 {
		printAns(arr, trace, trace[index].index)
	}
	fmt.Print(arr[index], " ")
}

func main() {
	var ans Value
	arr := input()
	
	fenwick_tree := make([]Value, MAX_NUM)
	trace := make([]Value, len(arr))
	for index, w := range arr {
		val := get(fenwick_tree, w-1)
		trace[index] = val

		new_val := Value{val.leng+1, index}

		update(fenwick_tree, w, new_val)

		if ans.leng < new_val.leng {
			ans = new_val
		}
	}
	
	printAns(arr, trace, ans.index)
	fmt.Println()
} 